File: Java/refactoring_exercise_2/src/test/java/za/co/entelect/refactoring2/controller/BankingControllerTest.java
package za.co.entelect.refactoring2.controller;

import org.junit.Test;
import za.co.entelect.refactoring2.domain.BankAccount;
import za.co.entelect.refactoring2.domain.BankingAction;
import za.co.entelect.refactoring2.domain.ChequeAccount;
import za.co.entelect.refactoring2.domain.SavingsAccount;
import za.co.entelect.refactoring2.exception.BankAccountException;
import za.co.entelect.refactoring2.service.ImageService;

import static junit.framework.Assert.*;

public class BankingControllerTest {

    public static final long INITIAL_BALANCE = 3000L;

    private BankingController bankingController = new BankingController();

    @Test
    public void testFetchImage(){
        assertNotNull(bankingController.fetchImage(ImageService.DEFAULT_IMAGE));
    }

    @Test
    public void testUploadImage(){
        assertNotNull(bankingController.uploadImage("1", new byte[1]));
    }

    @Test
    public void testRecalculateBalance(){
        SavingsAccount savingsAccount = createSavingsAccount(INITIAL_BALANCE);
        bankingController.updateAccount(savingsAccount, BankingAction.RECALCULATE_BALANCE);
        assertEquals(3150L, savingsAccount.getBalanceInCents());
    }

    @Test
    public void testCloseAccount(){
        SavingsAccount savingsAccount = createSavingsAccount(INITIAL_BALANCE);
        assertTrue(savingsAccount.isAccountActive());

        bankingController.updateAccount(savingsAccount, BankingAction.CLOSE_ACCOUNT);
        assertFalse(savingsAccount.isAccountActive());

        bankingController.updateAccount(savingsAccount, BankingAction.CLOSE_ACCOUNT);
        assertFalse(savingsAccount.isAccountActive());
    }

    @Test
    public void testReopenAccount(){
        SavingsAccount savingsAccount = createSavingsAccount(INITIAL_BALANCE);
        savingsAccount.closeAccount();

        assertFalse(savingsAccount.isAccountActive());
        bankingController.updateAccount(savingsAccount, BankingAction.REOPEN_ACCOUNT);

        assertTrue(savingsAccount.isAccountActive());
    }

    @Test
    public void testReopenOpenSavingsAccount(){
        SavingsAccount savingsAccount = createSavingsAccount(INITIAL_BALANCE);
        assertTrue(savingsAccount.isAccountActive());
        bankingController.updateAccount(savingsAccount, BankingAction.REOPEN_ACCOUNT);
        assertTrue(savingsAccount.isAccountActive());
    }

    @Test
    public void testReopenChequeAccount(){
        ChequeAccount chequeAccount = createChequeAccount(0);
        assertTrue(chequeAccount.isAccountActive());
        bankingController.updateAccount(chequeAccount, BankingAction.REOPEN_ACCOUNT);
        assertTrue(chequeAccount.isAccountActive());
    }

    @Test(expected = BankAccountException.class)
    public void testReopenAccountFail(){
        SavingsAccount savingsAccount = createSavingsAccount(0L);
        savingsAccount.closeAccount();
        bankingController.updateAccount(savingsAccount, BankingAction.REOPEN_ACCOUNT);
    }

    @Test
    public void testChargeAccountFee(){
        SavingsAccount savingsAccount = createSavingsAccount(INITIAL_BALANCE);
        bankingController.updateAccount(savingsAccount, BankingAction.CHARGE_ACCOUNT_FEE);
        assertEquals(savingsAccount.getBalanceInCents(), INITIAL_BALANCE - BankAccount.SAVINGS_ACCOUNT_FEE);
    }

    private SavingsAccount createSavingsAccount(long balance) {
        return new SavingsAccount(balance, BankAccount.SAVINGS_CREDIT_INTEREST_RATE, BankAccount.SAVINGS_DEBIT_INTEREST_RATE, BankAccount.SAVINGS_ACCOUNT_FEE);
    }

    private ChequeAccount createChequeAccount(long balance) {
        return new ChequeAccount(balance, BankAccount.CHEQUE_CREDIT_INTEREST_RATE, BankAccount.CHEQUE_DEBIT_INTEREST_RATE, BankAccount.CHEQUE_ACCOUNT_FEE);
    }
}

File: C#/Refactoring/refactoring_exercise_2/test/za/co/entelect/refactoring2/controller/BankingControllerTest.cs
using NUnit.Framework;
using NUnit.Framework.Constraints;
using refactoring_exercise_2.za.co.entelect.refactoring2.controller;
using refactoring_exercise_2.za.co.entelect.refactoring2.domain;
using refactoring_exercise_2.za.co.entelect.refactoring2.exception;
using refactoring_exercise_2.za.co.entelect.refactoring2.service;

namespace refactoring_exercise_2.test.za.co.entelect.refactoring2.controller
{
    [TestFixture]
    public class BankingControllerTest {

        public const long InitialBalance = 3000L;

        private readonly BankingController bankingController = new BankingController();

        [Test]
        public void TestFetchImage(){
            Assert.NotNull(bankingController.FetchImage(ImageService.DefaultImage));
        }

        [Test]
        public void TestUploadImage(){
            Assert.NotNull((bankingController.UploadImage("1", new byte[1])));
        }

        [Test]
        public void TestRecalculateBalance(){
            SavingsAccount savingsAccount = createSavingsAccount(InitialBalance);
            bankingController.UpdateAccount(savingsAccount, BankingAction.RecalculateBalance);
            Assert.That(savingsAccount.BalanceInCents, new EqualConstraint(3150L));
        }

        [Test]
        public void TestCloseAccount(){
            SavingsAccount savingsAccount = createSavingsAccount(InitialBalance);
            Assert.True(savingsAccount.IsAccountActive());

            bankingController.UpdateAccount(savingsAccount, BankingAction.CloseAccount);
            Assert.False(savingsAccount.IsAccountActive());

            bankingController.UpdateAccount(savingsAccount, BankingAction.CloseAccount);
            Assert.False(savingsAccount.IsAccountActive());
        }

        [Test]
        public void TestReopenAccount(){
            SavingsAccount savingsAccount = createSavingsAccount(InitialBalance);
            savingsAccount.CloseAccount();

            Assert.False(savingsAccount.IsAccountActive());
            bankingController.UpdateAccount(savingsAccount, BankingAction.ReopenAccount);

            Assert.True(savingsAccount.IsAccountActive());
        }

        [Test]
        public void TestReopenOpenSavingsAccount(){
            SavingsAccount savingsAccount = createSavingsAccount(InitialBalance);
            Assert.True(savingsAccount.IsAccountActive());
            bankingController.UpdateAccount(savingsAccount, BankingAction.ReopenAccount);
            Assert.True(savingsAccount.IsAccountActive());
        }

        [Test]
        public void TestReopenChequeAccount(){
            ChequeAccount chequeAccount = createChequeAccount(0);
            Assert.True(chequeAccount.IsAccountActive());
            bankingController.UpdateAccount(chequeAccount, BankingAction.ReopenAccount);
            Assert.True(chequeAccount.IsAccountActive());
        }

        [Test]
        [ExpectedException( typeof (BankAccountException))]
        public void TestReopenAccountFail(){
            SavingsAccount savingsAccount = createSavingsAccount(0L);
            savingsAccount.CloseAccount();
            bankingController.UpdateAccount(savingsAccount, BankingAction.ReopenAccount);
        }

        [Test]
        public void TestChargeAccountFee(){
            SavingsAccount savingsAccount = createSavingsAccount(InitialBalance);
            bankingController.UpdateAccount(savingsAccount, BankingAction.ChargeAccountFee);
            Assert.That(savingsAccount.BalanceInCents, new EqualConstraint(InitialBalance - BankAccount.SavingsAccountFee));
        }

        private SavingsAccount createSavingsAccount(long balance) {
            return new SavingsAccount(balance, BankAccount.SavingsCreditInterestRate, BankAccount.SavingsDebitInterestRate, BankAccount.SavingsAccountFee);
        }

        private ChequeAccount createChequeAccount(long balance) {
            return new ChequeAccount(balance, BankAccount.ChequeCreditInterestRate, BankAccount.ChequeDebitInterestRate, BankAccount.ChequeAccountFee);
        }
    }
}